### PR TITLE
Add TypeScript support

### PR DESCRIPTION
**Pull request by CloudNiner**
Via manual types provided by index.d.ts.

See discussion in https://github.com/azavea/react-showtime/pull/10
for the alternatives discussed.

The types are compiled in CI, which should
catch at least some potential issues with
API changes.

This PR supersedes #10, somehow the git history got corrupted and between that and a nasty rebase it seemed easier to just start fresh for a small independent changeset.

### CODE REVIEW

**Comment by pcaisse** [comment]
In looking at the deploy preview, it looks like this breaks the examples. I think we might just need to rename the file to `TSExample.tsx` or something similar since the import `import Example from "./Example";` is ambiguous as to which file to use and it's choosing the `.tsx` one.

File: example/src/components/Example.tsx
Comment (pcaisse) [inline example/src/components/Example.tsx:5]: Is there an easy was for us to access this for our own manual testing purposes but not show it as an official example as a little TypeScript easter egg? I think that would be ideal if we could, though not required. It's a little weird as is because I can't tell if the example actually works, only that it type checks.

Comment (pcaisse) [reply example/src/components/Example.tsx]: > Perhaps we could get what you're suggesting just by moving our Example.tsx to the src dir since it's not really a public example yet?
Yeah, that's probably good for now.

Comment (CloudNiner) [reply example/src/components/Example.tsx]: Updated. Go ahead and resolve this thread if sufficient.

**Review by pcaisse — APPROVED** [review]
LGTM!

**Comment by CloudNiner** [comment]
@lederer did you want to take a look at this before we merge?

**Comment by lederer** [comment]
Yup, I'll look today. Thanks!

File: index.d.ts
Comment (lederer) [inline index.d.ts:44]: Should this be `settings?:`, since it's possible to call `useShowtime()` with no params?

File: index.d.ts
Comment (lederer) [inline index.d.ts:9]: Strictly speaking, this might be better named as singular `TransitionProperty` since it represents a single CSS property that will be transitioned. That also matched the "CSS Property" term used to describe it in the API section of the Readme.

**Review by lederer — CHANGES_REQUESTED** [review]
Something seems off with placing `TypeScriptExample.tsx` directly under `src/`. All (non-dot) files in the top level of that folder are directly used by the library in production, whereas `TypeScriptExample.tsx` is not imported or referenced anywhere and is intended as an instructive example of a use case.

It also doesn't really belong under the `example` folder. The name of that folder is perhaps a mistake, as it's really more of a demo. IIRC the folder was generated by `create-react-hook`, the library I used to bootstrap this project, and I didn't think twice about it.

I'm not sure what the right answer is here, but some ideas are:
- Place it in the top level of the repo. This would also make it more discoverable to TS consumers.
- Place it in a new folder at the top level of the repo. Something like `TypeScript/` or `TypeScriptExample/`

File: index.d.ts
Comment (lederer) [inline index.d.ts:5]: I don't know if this matters from a TS perspective, but should `Showtime` here (and elsewhere in this file) be changed to `ReactShowtime`, since that's the library's full name? In the context of the library `Showtime` generally refers to the `<Showtime>` component rather than the library.

Comment (CloudNiner) [reply index.d.ts:5]: @pcaisse I defer to you on this one, I believe this was a change you added.

**Comment by CloudNiner** [comment]
> I'm not sure what the right answer is here, but some ideas

Well summarized. Agree it wasn't quite right.

Perhaps `example-ts/Example.tsx` or `ts/Example.tsx` to keep the lower case folder naming convention consistent? Also doesn't feel quite right to clutter the root directory? Someone looking for typescript should be able to match the `ts` abbreviation and know what `tsx` files are.

**Comment by lederer** [comment]
Makes sense. I vote `ts/Example.tsx`

File: index.d.ts
Comment (pcaisse) [reply index.d.ts:5]: Did I? I thought you added it :joy: I'm not sure if this is needed since all the types will live in this one file.

Comment (CloudNiner) [reply index.d.ts:5]: Oh jeez. Also not confident its necessary. Let's remove for now?

Comment (pcaisse) [reply index.d.ts:5]: Yeah, if the build passes without it let's ditch it.

Comment (pcaisse) [reply index.d.ts:5]: Ah, you edited your comment! :detective: I don't really understand the use case for namespaces unless you're using defining types across files, but then again I don't know that much about namespaces.

Here's another JS-first React project that we use that had TS types added after the fact and it doesn't use a namespace FWIW:
https://github.com/jbetancur/react-data-table-component/blob/master/index.d.ts

I'm fine with renaming it or deleting it.

Comment (CloudNiner) [reply index.d.ts:5]: Yeah sorry for the ninja edit. That's always a bad idea. Good enough of an example for me. Removing.

**Comment by CloudNiner** [comment]
@lederer updated -- all comments addressed.

**Review by lederer — APPROVED** [review]

**Comment by lederer** [comment]
Thanks @CloudNiner for working through the changes. And thanks to you and @pcaisse for diving in to broaden our audience. Really appreciate!

**Pull request merged**
